File: etna/lib/etna/directed_graph.rb
class DirectedGraph
  def initialize
    @children = {}
    @parents = {}
  end

  attr_reader :children
  attr_reader :parents

  def full_parentage(n)
    [].tap do |result|
      q = @parents[n].keys.dup
      seen = Set.new

      until q.empty?
        n = q.shift
        next if seen.include?(n)
        seen.add(n)

        result << n
        q.push(*@parents[n].keys)
      end

      result.uniq!
    end
  end

  def as_normalized_hash(root, include_root = true)
    q = [root]
    {}.tap do |result|
      if include_root
        result[root] = []
      end

      seen = Set.new

      until q.empty?
        n = q.shift
        next if seen.include?(n)
        seen.add(n)

        parentage = full_parentage(n)

        @children[n].keys.each do |child_node|
          q << child_node

          if result.include?(n)
            result[n] << child_node
          end

          parentage.each do |grandparent|
            result[grandparent] << child_node if result.include?(grandparent)
          end

          # Depending on the graph shape, diamonds could lead to
          #   resetting of previously calculated dependencies.
          # Here we avoid resetting existing entries in `result`
          #   and instead concatenate them if they already exist.
          result[child_node] = [] unless result.include?(child_node)
          result[n].concat(result[child_node]) if result.include?(child_node) && result.include?(n)
        end
      end

      result.values.each(&:uniq!)
    end
  end

  def add_connection(parent, child)
    children = @children[parent] ||= {}
    child_children = @children[child] ||= {}

    children[child] = child_children

    parents = @parents[child] ||= {}
    parent_parents = @parents[parent] ||= {}
    parents[parent] = parent_parents
  end

  def serialized_path_from(root, include_root = true)
    seen = Set.new
    [].tap do |result|
      result << root if include_root
      seen.add(root)
      path_q = paths_from(root, include_root)
      traversables = path_q.flatten

      until path_q.empty?
        next_path = path_q.shift
        next if next_path.nil?

        until next_path.empty?
          next_n = next_path.shift
          next if next_n.nil?
          next if seen.include?(next_n)

          if @parents[next_n].keys.any? { |p| !seen.include?(p) && traversables.include?(p) }
            next_path.unshift(next_n)
            path_q.push(next_path)
            break
          else
            result << next_n
            seen.add(next_n)
          end
        end
      end
    end
  end

  def paths_from(root, include_root = true)
    [].tap do |result|
      parents_of_map = descendants(root, include_root)
      seen = Set.new

      parents_of_map.to_a.sort_by { |k, parents| [-parents.length, k.inspect] }.each do |k, parents|
        unless seen.include?(k)
          if @children[k].keys.empty?
            result << parents + [k]
          else
            @children[k].keys.dup.sort.each do |c|
              result << parents + [k, c]
            end
          end
        end

        parents.each { |p| seen.add(p) }
      end
    end
  end

  def descendants(parent, include_root = true)
    seen = Set.new

    seen.add(parent)
    queue = @children[parent].keys.dup
    parent_queue = @parents[parent].keys.dup

    # Because this is not an acyclical graph, the definition of descendants needs to be stronger;
    # here we believe that any path that would move through --any-- parent to this child would not be considered
    # descendant, so we first find all those parents and mark them as 'seen' so that they are not traveled.
    while next_parent = parent_queue.pop
      next if seen.include?(next_parent)
      seen.add(next_parent)
      parent_queue.push(*@parents[next_parent].keys)
    end

    queue = queue.nil? ? [] : queue.dup
    paths = {}

    while child = queue.pop
      next if seen.include? child
      seen.add(child)
      path = (paths[child] ||= (include_root ? [parent] : []))

      @children[child].keys.each do |child_child|
        queue.push child_child

        unless paths.include? child_child
          paths[child_child] = path + [child]
        end
      end
    end

    paths
  end
end

File: etna/packages/etna-js/utils/directed_graph.js
"use strict";
// A port of directed_graph.rb to TypeScript
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
};
exports.__esModule = true;
exports.DirectedGraph = void 0;
var DirectedGraph = /** @class */ (function () {
    function DirectedGraph() {
        this.children = {};
        this.parents = {};
    }
    DirectedGraph.prototype.fullParentage = function (n) {
        var result = [];
        if (!this.parents[n])
            return result;
        var q = Object.keys(this.parents[n]);
        var seen = new Set();
        while (q.length > 0) {
            var next = q.splice(0, 1)[0];
            if (seen.has(next))
                continue;
            seen.add(next);
            result.push(next);
            q = q.concat(Object.keys(this.parents[next]));
        }
        return __spreadArray([], __read(new Set(result)));
    };
    DirectedGraph.prototype.asNormalizedHash = function (root, includeRoot) {
        if (includeRoot === void 0) { includeRoot = true; }
        var q = [root];
        var result = {};
        if (includeRoot)
            result[root] = [];
        var seen = new Set();
        var _loop_1 = function () {
            var next = q.splice(0, 1)[0];
            if (seen.has(next))
                return "continue";
            seen.add(next);
            var parentage = this_1.fullParentage(next);
            Object.keys(this_1.children[next]).forEach(function (childNode) {
                q.push(childNode);
                if (result.hasOwnProperty(next))
                    result[next].push(childNode);
                parentage.forEach(function (grandparent) {
                    if (result.hasOwnProperty(grandparent))
                        result[grandparent].push(childNode);
                });
                // Depending on the graph shape, diamonds could lead to
                //    resetting of previously calculated dependencies.
                // Here we avoid resetting existing entries in `result`
                //    and instead concatenate them if they already exist.
                if (!result.hasOwnProperty(childNode))
                    result[childNode] = [];
                if (result.hasOwnProperty(childNode) && result.hasOwnProperty(next))
                    result[next] = result[next].concat(result[childNode]);
            });
        };
        var this_1 = this;
        while (q.length > 0) {
            _loop_1();
        }
        return Object.entries(result).reduce(function (acc, _a) {
            var _b = __read(_a, 2), key = _b[0], value = _b[1];
            acc[key] = __spreadArray([], __read(new Set(value)));
            return acc;
        }, {});
    };
    DirectedGraph.prototype.addConnection = function (parent, child) {
        if (!this.children.hasOwnProperty(parent)) {
            this.children[parent] = {};
        }
        if (!this.children.hasOwnProperty(child)) {
            this.children[child] = {};
        }
        this.children[parent][child] = this.children[child];
        if (!this.parents.hasOwnProperty(child)) {
            this.parents[child] = {};
        }
        if (!this.parents.hasOwnProperty(parent)) {
            this.parents[parent] = {};
        }
        this.parents[child][parent] = this.parents[parent];
    };
    DirectedGraph.prototype.serializedPathFrom = function (root, includeRoot) {
        if (includeRoot === void 0) { includeRoot = true; }
        var seen = new Set();
        var result = [];
        if (includeRoot)
            result.push(root);
        seen.add(root);
        var pathQ = this.pathsFrom(root, includeRoot);
        var traversables = pathQ.flat();
        while (pathQ.length > 0) {
            var nextPath = pathQ.splice(0, 1)[0];
            if (!nextPath || nextPath.length === 0)
                continue;
            while (nextPath.length > 0) {
                var nextN = nextPath.splice(0, 1)[0];
                if (!nextN)
                    continue;
                if (seen.has(nextN))
                    continue;
                if (Object.keys(this.parents[nextN]).some(function (p) {
                    return !seen.has(p) && traversables.includes(p);
                })) {
                    nextPath.unshift(nextN);
                    pathQ.push(nextPath);
                    break;
                }
                else {
                    result.push(nextN);
                    seen.add(nextN);
                }
            }
        }
        return result;
    };
    DirectedGraph.prototype.pathsFrom = function (root, includeRoot) {
        var _this = this;
        if (includeRoot === void 0) { includeRoot = true; }
        var result = [];
        var parentsOfMap = this.descendants(root, includeRoot);
        var seen = new Set();
        Object.entries(parentsOfMap)
            .sort(function (_a, _b) {
            var _c = __read(_a, 2), k1 = _c[0], p1 = _c[1];
            var _d = __read(_b, 2), k2 = _d[0], p2 = _d[1];
            // if p1 is longer, sort it before p2
            if (p1.length !== p2.length)
                return p1.length > p2.length ? -1 : 1;
            return k1 > k2 ? 1 : -1;
        })
            .forEach(function (_a) {
            var _b = __read(_a, 2), key = _b[0], parents = _b[1];
            if (!seen.has(key)) {
                if (Object.keys(_this.children[key]).length === 0) {
                    result.push(parents.concat([key]));
                }
                else {
                    Object.keys(_this.children[key])
                        .sort()
                        .forEach(function (c) {
                        result.push(parents.concat([key, c]));
                    });
                }
            }
            parents.forEach(function (p) { return seen.add(p); });
        });
        return result;
    };
    DirectedGraph.prototype.descendants = function (parent, includeRoot) {
        if (includeRoot === void 0) { includeRoot = true; }
        var seen = new Set();
        seen.add(parent);
        var q = Object.keys(this.children[parent]);
        var parentQ = Object.keys(this.parents[parent]);
        // Because this is not an acyclical graph, the definition of descendants needs to be stronger;
        // here we believe that any path that would move through --any-- parent to this child would not be considered
        // descendant, so we first find all those parents and mark them as 'seen' so that they are not traveled.
        while (parentQ.length > 0) {
            var nextParent = parentQ.pop();
            if (!nextParent)
                break;
            if (seen.has(nextParent))
                continue;
            seen.add(nextParent);
            parentQ = parentQ.concat(Object.keys(this.parents[nextParent]));
        }
        var paths = {};
        var _loop_2 = function () {
            var child = q.pop();
            if (!child)
                return "break";
            if (seen.has(child))
                return "continue";
            seen.add(child);
            if (!paths.hasOwnProperty(child))
                paths[child] = includeRoot ? [parent] : [];
            var path = paths[child];
            Object.keys(this_2.children[child]).forEach(function (childChild) {
                q.push(childChild);
                if (!paths.hasOwnProperty(childChild))
                    paths[childChild] = path.concat([child]);
            });
        };
        var this_2 = this;
        while (q.length > 0) {
            var state_1 = _loop_2();
            if (state_1 === "break")
                break;
        }
        return paths;
    };
    return DirectedGraph;
}());
exports.DirectedGraph = DirectedGraph;
